src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,796 @@
+    mAdvertisedOnLinkPrefix = newOnLinkPrefix;
+    static_assert(sizeof(mAdvertisedOmrPrefixes) == sizeof(newOmrPrefixes), "invalid new OMR prefix array size");
+    memcpy(mAdvertisedOmrPrefixes, newOmrPrefixes, sizeof(newOmrPrefixes[0]) * newOmrPrefixNum);
+    mAdvertisedOmrPrefixNum = newOmrPrefixNum;
+
+exit:
+    return;
+}
+
+void RoutingManager::StartRouterSolicitation(void)
+{
+    mRouterSolicitCount = 0;
+    mRouterSolicitTimer.Start(Random::NonCrypto::GetUint32InRange(0, kMaxRtrSolicitationDelay * 1000));
+}
+
+otError RoutingManager::SendRouterSolicitation(void)
+{
+    Ip6::Address                    destAddress;
+    RouterAdv::RouterSolicitMessage routerSolicit;
+
+    OT_ASSERT(IsInitialized());
+
+    destAddress.SetToLinkLocalAllRoutersMulticast();
+    return otPlatInfraIfSendIcmp6(mInfraIfIndex, &destAddress, reinterpret_cast<const uint8_t *>(&routerSolicit),
+                                  sizeof(routerSolicit));
+}
+
+void RoutingManager::SendRouterAdvertisement(const Ip6::Prefix *aNewOmrPrefixes,
+                                             uint8_t            aNewOmrPrefixNum,
+                                             const Ip6::Prefix *aNewOnLinkPrefix)
+{
+    uint8_t                     buffer[kMaxRouterAdvMessageLength];
+    uint16_t                    bufferLength = 0;
+    RouterAdv::RouterAdvMessage routerAdv;
+
+    // Set zero Router Lifetime to indicate that the Border Router is not the default
+    // router for infra link so that hosts on infra link will not create default route
+    // to the Border Router when received RA.
+    routerAdv.SetRouterLifetime(0);
+
+    OT_ASSERT(bufferLength + sizeof(routerAdv) <= sizeof(buffer));
+    memcpy(buffer, &routerAdv, sizeof(routerAdv));
+    bufferLength += sizeof(routerAdv);
+
+    if (aNewOnLinkPrefix != nullptr)
+    {
+        RouterAdv::PrefixInfoOption pio;
+
+        pio.SetOnLink(true);
+        pio.SetAutoAddrConfig(true);
+        pio.SetValidLifetime(kDefaultOnLinkPrefixLifetime);
+        pio.SetPreferredLifetime(kDefaultOnLinkPrefixLifetime);
+        pio.SetPrefix(*aNewOnLinkPrefix);
+
+        OT_ASSERT(bufferLength + pio.GetLength() <= sizeof(buffer));
+        memcpy(buffer + bufferLength, &pio, pio.GetLength());
+        bufferLength += pio.GetLength();
+
+        if (mAdvertisedOnLinkPrefix != nullptr)
+        {
+            otLogInfoBr("start advertising new on-link prefix %s on interface %u",
+                        aNewOnLinkPrefix->ToString().AsCString(), mInfraIfIndex);
+        }
+
+        otLogInfoBr("send on-link prefix %s in PIO (valid lifetime = %u seconds)",
+                    aNewOnLinkPrefix->ToString().AsCString(), kDefaultOnLinkPrefixLifetime);
+    }
+    else if (mAdvertisedOnLinkPrefix != nullptr)
+    {
+        RouterAdv::PrefixInfoOption pio;
+
+        // Set zero valid lifetime to immediately invalidate the advertised on-link prefix.
+        pio.SetValidLifetime(0);
+        pio.SetPreferredLifetime(0);
```

Comment:
yes, the code here is just for readability. I think it will be easier for a reader to get the idea of removing a prefix by setting its valid lifetime to 0.